packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java

```diff
@@ -0,0 +1,248 @@
+      int requestCode, String[] permissions, int[] grantResults) {
+    if (requestCode == REQUEST_IMAGE_PICKER_PERMISSIONS) {
+      boolean bothPermissionsGranted =
+          grantResults[0] == PackageManager.PERMISSION_GRANTED
+              && grantResults[1] == PackageManager.PERMISSION_GRANTED;
+
+      if (bothPermissionsGranted) {
+        executePendingOperation();
+      } else {
+        pendingResult.error(
+            "no_permissions", "image_picker plugin requires camera permissions", null);
+        clearMethodCallAndResult();
+      }
+      return true;
+    }
+    return false;
+  }
+
+  private void executePendingOperation() {
+    if (pendingPickImageOperation != null) {
+      switch (pendingPickImageOperation) {
+        case CHOOSE_FROM_GALLERY:
+          launchPickImageFromGalleryIntent();
+          break;
+        case TAKE_WITH_CAMERA:
+          launchTakeImageWithCameraIntent();
+          break;
+      }
+
+      pendingPickImageOperation = null;
+    }
+  }
+
+  @Override
+  public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
+    if (requestCode == REQUEST_CODE_CHOOSE_FROM_GALLERY) {
+      return handleChoosePictureResult(resultCode, data);
+    } else if (requestCode == REQUEST_CODE_TAKE_WITH_CAMERA) {
+      return handleTakePictureResult(resultCode);
+    }
+    return false;
+  }
+
+  private boolean handleChoosePictureResult(int resultCode, Intent data) {
+    if (resultCode == Activity.RESULT_OK && data != null) {
+      String path = FileUtils.getPathFromUri(registrar, data.getData());
+      handleResult(path);
+      return true;
+    } else if (resultCode != Activity.RESULT_CANCELED) {
+      pendingResult.error("PICK_ERROR", "Error picking image", null);
+    }
+
+    clearMethodCallAndResult();
+    return true;
+  }
+
+  private boolean handleTakePictureResult(int resultCode) {
+    if (resultCode == Activity.RESULT_OK) {
+      MediaScannerConnection.scanFile(
+          registrar,
+          new String[] {pendingCameraImageUri.getPath()},
+          null,
+          new MediaScannerConnection.OnScanCompletedListener() {
+            @Override
+            public void onScanCompleted(String path, Uri uri) {
+              handleResult(path);
+            }
+          });
+      return true;
+    } else if (resultCode != Activity.RESULT_CANCELED) {
+      pendingResult.error("PICK_ERROR", "Error taking photo", null);
+      clearMethodCallAndResult();
+    }
+    return true;
+  }
+
+  private void handleResult(String path) {
+    if (pendingResult != null) {
+      Double maxWidth = methodCall.argument("maxWidth");
+      Double maxHeight = methodCall.argument("maxHeight");
+
+      String finalImagePath = imageResizer.resizeImageIfNeeded(path, maxWidth, maxHeight);
+      pendingResult.success(finalImagePath);
+
+      clearMethodCallAndResult();
+    } else {
+      throw new IllegalStateException("Received images from picker that were not requested");
```

Comment:
@mravn-google When should we throw and when return an error from a plugin?

Reply:
I think we should use `result.error(...)` explicitly only for errors that are relevant to the abstraction provided by the channel-based API. Programming errors in Java code should just lead to RuntimeExceptions, as usual. If the occur as part of `onMethodCall`, they would be logged on the Java side, and relayed to the Dart side ([code](https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/plugin/common/MethodChannel.java#L208-L209)).